add uv.lock support

cataggar opened the pull request:
This is an attempt to get uv support started by adding a UvLockComponentDetector.
- #1372

Hopefully, I am on the right track. Let me know.

FernandoRojo commented on src/Microsoft.ComponentDetection.Detectors/uv/UvLockComponentDetector.cs:
```diff
@@ -0,0 +1,76 @@
+    using Microsoft.ComponentDetection.Contracts.TypedComponent;
+    using Microsoft.Extensions.Logging;
+
+    public class UvLockComponentDetector : FileComponentDetector
```
We'll want to make sure to initially have this with the DefaultOffComponentDetector interface to test prior to enabling in prod

cataggar replied:
https://github.com/microsoft/component-detection/blob/main/docs/creating-a-new-detector.md#detector-lifecycle

grvillic commented on docs/detectors/uv.md:
```diff
@@ -0,0 +1,6 @@
+## Requirements
+uv detection relies on a uv.lock file being present.
+
+## Detection strategy
```
Could we place links here to `uv` official documentation?

cataggar replied:
Links added.

grvillic commented on src/Microsoft.ComponentDetection.Orchestrator/Services/DetectorRestrictionService.cs:
```diff
@@ -20,26 +20,33 @@ public IEnumerable<IComponentDetector> ApplyRestrictions(DetectorRestrictions re
     {
```
Do not make any changes to this file, at least not in this PR. We are able to enable disabled detectors with the arguments, default scan doesn't have any `DetectorsFilter`. We can do a follow up PR just for this.

grvillic commented:
Double check the tests, they seem to be failing in some environments.

cataggar commented:
@grvillic, I fixed the unit tests and removed the `DetectorRestrictionService` changes, but how to I run just the `UvLockDetector` locally?

```
PS C:\Users\cataggar\ms\component-detection> dotnet run --project ./src/Microsoft.ComponentDetection/Microsoft.ComponentDetection.csproj scan --DetectorsFilter UvLock --DetectorArgs UvLock=EnableIfDefaultOff --SourceDirectory $sourceDir
[13:05:57 ERR] An error occurred while executing the command
Microsoft.ComponentDetection.Orchestrator.Exceptions.InvalidDetectorFilterException: Detector 'UvLock' was not found
   at Microsoft.ComponentDetection.Orchestrator.Services.DetectorRestrictionService.ApplyRestrictions(DetectorRestrictions restrictions, IEnumerable`1 detectors) in C:\Users\cataggar\ms\component-detection\src\Microsoft.ComponentDetection.Orchestrator\Services\DetectorRestrictionService.cs:line 46
   at Microsoft.ComponentDetection.Orchestrator.Services.ScanExecutionService.ExecuteScanAsync(ScanSettings settings) in C:\Users\cataggar\ms\component-detection\src\Microsoft.ComponentDetection.Orchestrator\Services\ScanExecutionService.cs:line 42
   at Microsoft.ComponentDetection.Orchestrator.Commands.ScanCommand.ExecuteAsync(CommandContext context, ScanSettings settings) in C:\Users\cataggar\ms\component-detection\src\Microsoft.ComponentDetection.Orchestrator\Commands\ScanCommand.cs:line 43
   at Spectre.Console.Cli.CommandExecutor.Execute(CommandTree leaf, CommandTree tree, CommandContext context, ITypeResolver resolver, IConfiguration configuration) in /_/src/Spectre.Console.Cli/Internal/CommandExecutor.cs:line 166
```

grvillic commented on src/Microsoft.ComponentDetection.Orchestrator/Experiments/Configs/UvLockDetectorExperiment.cs:
```diff
@@ -0,0 +1,23 @@
+    public string Name => "UvLockDetectorExperiment";
+
+    /// <inheritdoc />
+    public bool IsInControlGroup(IComponentDetector componentDetector) => componentDetector is PipComponentDetector;
```
Let's compare against `PipReportComponentDetector` that is the compliant detector used when Network Isolation is enforced.

cataggar commented on src/Microsoft.ComponentDetection.Orchestrator/Experiments/Configs/UvLockDetectorExperiment.cs:
```diff
@@ -0,0 +1,23 @@
+/// <summary>
+/// Experiment to validate UvLockComponentDetector against PipComponentDetector.
+/// </summary>
+public class UvLockDetectorExperiment : IExperimentConfiguration
```
Looks like this should have been `DefaultOffComponentDetector` to begin with https://github.com/microsoft/component-detection/blob/main/docs/creating-a-new-detector.md#detector-lifecycle.
